spec/lib/index_spec.rb

```diff
@@ -0,0 +1,141 @@
+
+token = "eyJraWQiOiJEYVgxMWdBcldRZWJOSE83RU1QTUw1VnRUNEV3cmZrd2M1U2xHaVd2VXdBIiwiYWxnIjoiUlMyNTYifQ.eyJzdWIiOiIwMHVkanl\
+  qc3NidDJTMVFWcjBoNyIsInZlciI6MSwiaXNzIjoiaHR0cHM6Ly9hdXRoZW50aWMuYXJ0aWN1bGF0ZS5jb20vIiwiYXVkIjoiMG9hZGp5azUyM2hsWmZ5\
+  YjEwaDciLCJpYXQiOjE1MTY2MzcwOTEsImV4cCI6MTUxNjY0MDY5MSwianRpIjoiSUQuYzhqaHpva3kwZkZOUHI4TF9TQ3JycGdUVFV5QW9jdEh2NjlPS\
+  1NtZjVHQSIsImFtciI6WyJwd2QiXSwiaWRwIjoiMDBvY2c0dGJ1NkZLMkRoNUcwaDciLCJub25jZSI6IjIiLCJhdXRoX3RpbWUiOjE1MTY2MzcwOTEsIn\
+  RlbmFudElkIjoiZDQyZTMzZmQtZjA1ZS00YTRlLTkwNTAtNWI3YjJlODAwODM0In0.NEVqz-jJIyaEgho3uQYOvWC52s_50AV--FHwBWm9BftucQ5G4bS\
+  HL7szeaPc3HT0VrhFUntRLlJHzw7pZvRJG2WExj6HJi-Ug3LDwQOj47Gf_ywlEydBAQz7u98JK2ZJcCP16-lIOM1J-fUz-SpFqI4RcO5MLiiEPnMqsXS-\
+  EkPd8Y27G64PnHnNjaY3sLrOc9peeD5Xh82TSjeMFFAPpiYNtTCixnfZeQCCtxOCPhiDYAwDSxaLbrOcDAYdO0ytKQ9dBfFoY0AzJNqgJUOPVeeC_AgEJ\
+  eLIaSKVJAFqZAB8t5VagvVGIqcu7TaMCOmOZx_5A8Xc9JVmRoKDAMlizQ"
+
+bad_token = "cyJraWQiOiJEYVgxMWdBcldRZWJOSE83RU1QTUw1VnRUNEV3cmZrd2M1U2xHaVd2VXdBIiwiYWxnIjoiUlMyNTYifQ.eyJzdWIiOiIwMHV\
+  kanlqc3NidDJTMVFWcjBoNyIsInZlciI6MSwiaXNzIjoiaHR0cHM6Ly9hdXRoZW50aWMuYXJ0aWN1bGF0ZS5jb20vIiwiYXVkIjoiMG9hZGp5azUyM2hs\
+  WmZ5YjEwaDciLCJpYXQiOjE1MTY2MzcwOTEsImV4cCI6MTUxNjY0MDY5MSwianRpIjoiSUQuYzhqaHpva3kwZkZOUHI4TF9TQ3JycGdUVFV5QW9jdEh2N\
+  jlPS1NtZjVHQSIsImFtciI6WyJwd2QiXSwiaWRwIjoiMDBvY2c0dGJ1NkZLMkRoNUcwaDciLCJub25jZSI6IjIiLCJhdXRoX3RpbWUiOjE1MTY2MzcwOT\
+  EsInRlbmFudElkIjoiZDQyZTMzZmQtZjA1ZS00YTRlLTkwNTAtNWI3YjJlODAwODM0In0.NEVqz-jJIyaEgho3uQYOvWC52s_50AV--FHwBWm9BftucQ5\
+  G4bSHL7szeaPc3HT0VrhFUntRLlJHzw7pZvRJG2WExj6HJi-Ug3LDwQOj47Gf_ywlEydBAQz7u98JK2ZJcCP16-lIOM1J-fUz-SpFqI4RcO5MLiiEPnMq\
+  sXS-EkPd8Y27G64PnHnNjaY3sLrOc9peeD5Xh82TSjeMFFAPpiYNtTCixnfZeQCCtxOCPhiDYAwDSxaLbrOcDAYdO0ytKQ9dBfFoY0AzJNqgJUOPVeeC_\
+  AgEJeLIaSKVJAFqZAB8t5VagvVGIqcu7TaMCOmOZx_5A8Xc9JVmRoKDAMlizQ"
+
+describe 'Authentic' do
+  before do
+    oidc_file = File.read('./spec/fixtures/oidc.json')
+    key_file = File.read('./spec/fixtures/keys.json')
+    @oidc = JSON.parse(oidc_file)
+    stub_request(:get, test_url).to_return(body: oidc_file)
+    stub_request(:get, test_jwks_url).to_return(body: key_file)
+  end
+
+  before(:each) do
+    opts = { issWhiteList: [@oidc['issuer']], cacheMaxAge: '1m' }
+    @test_instance = Authentic::Validator.new opts
+  end
+
+  describe 'init class' do
+    it 'errors if no opts are provided' do
+      opts = {}
+      expect { Authentic::Validator.new opts }.to raise_error(Authentic::IncompleteOptions)
+    end
+
+    it 'errors if no issWhiteList urls are provided' do
+      opts = { issWhiteList: [] }
+      expect { Authentic::Validator.new opts }.to raise_error(Authentic::IncompleteOptions)
+    end
+  end
+
+  describe '.valid' do
+    it 'validates the jwt against the jwks' do
+      expect(@test_instance.valid(token)).to be(true)
+    end
+
+    it 'caches the jwks client' do
+      # hydrates caches
+      expect(@test_instance.valid(token)).to be(true)
```

Comment:
Yea good point. Just me being lazy there

Reply:
Hmm interesting. Definitely possible and I agree with the sentiment but I personally like flow/integration tests. Maybe the problem is this is an integration test among unit tests. Otherwise I have to repeat this logic, via a method of course, and have each test assert one piece of the flow. I would almost rather add a message to the assertion to make it really obvious.

Reply:
Yeah, it's a tricky one, but is often worth decomposing for better test output when things break. In this case if it's a really complex flow then it may not be worth decomposing.

Reply:
yea makes sense. I think this one is simple enough to follow that pattern.